#!/usr/bin/env bash

#PBS -l nodes=1:ppn=12
#PBS -l mem=48gb
#PBS -l walltime=24:00:00
#PBS -q short
#PBS -t 1
#PBS -N STAR-map
#PBS -j oe
#PBS -o log/$PBS_JOBNAME

cd $PBS_O_WORKDIR

FQLIST=../variant_calling/fastq_list.txt

SAMPLE=$( awk -v ARRID="$PBS_ARRAYID" 'FNR==ARRID { print $1 }' $FQLIST )
FQ1=$( awk -v ARRID="$PBS_ARRAYID" 'FNR==ARRID { print $2 }' $FQLIST )
FQ2=$( awk -v ARRID="$PBS_ARRAYID" 'FNR==ARRID { print $3 }' $FQLIST )
STAR=$HOME/HLAMAPPER/STAR-2.7.3a/bin/Linux_x86_64_static/STAR
INDEX=../../indices/STAR_v3.7.3a
OUT=/scratch/vitor/results/star/one_pass/${SAMPLE}

$STAR --runMode alignReads \
    --runThreadN $PBS_NUM_PPN \
    --genomeDir $INDEX \
    --readFilesIn $FQ1 $FQ2 \
    --readFilesCommand zcat  \
    --outFilterMismatchNmax 999 \
    --outFilterMismatchNoverReadLmax 0.04 \
    --outFilterMultimapNmax 20 \
    --outSAMtype BAM SortedByCoordinate \
    --outSAMunmapped Within KeepPairs \
    --outFileNamePrefix ${OUT}_

samtools index ${OUT}_Aligned.sortedByCoord.out.bam
